OpenRA.Mods.Common/Traits/Buildings/DockClientManager.cs

```diff
@@ -0,0 +1,327 @@
+			if (host.Reserve(hostActor, this))
+			{
+				ReservedHost = host;
+				ReservedHostActor = hostActor;
+
+				// After we have reserved a new Host we want to forget our old host.
+				lastReservedDockHost = null;
+				return true;
+			}
+
+			return false;
+		}
+
+		public void DockStarted(Actor self, Actor hostActor, DockHost host)
+		{
+			foreach (var client in dockClients)
+				client.DockStarted(self, hostActor, host);
+		}
+
+		public bool DockTick(Actor self, Actor hostActor, DockHost host)
+		{
+			if (IsTraitDisabled)
+				return true;
+
+			var cancel = true;
+			foreach (var client in dockClients)
+				if (!client.DockTick(self, hostActor, host))
+					cancel = false;
+
+			return cancel;
+		}
+
+		public void DockCompleted(Actor self, Actor hostActor, DockHost host)
+		{
+			foreach (var client in dockClients)
+				client.DockCompleted(self, hostActor, host);
+
+			UnreserveHost();
+		}
+
+		IEnumerable<IOrderTargeter> IIssueOrder.Orders
+		{
+			get
+			{
+				yield return new EnterAlliedActorTargeter<DockHostInfo>(
+					"ForceDock",
+					6,
+					Info.EnterCursor,
+					Info.EnterBlockedCursor,
+					DockingPossible,
+					target => CanDockAt(target, true, true));
+				yield return new EnterAlliedActorTargeter<DockHostInfo>(
+					"Dock",
+					5,
+					Info.EnterCursor,
+					Info.EnterBlockedCursor,
+					(actor, modifiers) => DockingPossible(actor),
+					target => CanDockAt(target, false, true));
+			}
+		}
+
+		void IResolveOrder.ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "Dock")
+			{
+				var target = order.Target;
+
+				// Deliver orders are only valid for own/allied actors,
+				// which are guaranteed to never be frozen.
+				// TODO: support frozen actors
+				if (target.Type != TargetType.Actor)
+					return;
+
+				if (IsTraitDisabled)
+					return;
+
+				var dock = AvailableDockHosts(target.Actor, false, true).ClosestDock(self, this);
+				if (!dock.HasValue)
+					return;
+
+				self.QueueActivity(order.Queued, new MoveToDock(self, dock.Value.Actor, dock.Value.Trait));
+				self.ShowTargetLines();
+			}
+			else if (order.OrderString == "ForceDock")
+			{
+				var target = order.Target;
+
+				// Deliver orders are only valid for own/allied actors,
+				// which are guaranteed to never be frozen.
+				// TODO: support frozen actors
+				if (target.Type != TargetType.Actor)
+					return;
+
+				if (IsTraitDisabled)
+					return;
+
+				var dock = AvailableDockHosts(target.Actor, true, true).ClosestDock(self, this);
+				if (!dock.HasValue)
+					return;
+
+				self.QueueActivity(order.Queued, new DeliverResources(self, dock.Value.Actor, dock.Value.Trait));
+				self.ShowTargetLines();
+			}
+		}
+
+		string IOrderVoice.VoicePhraseForOrder(Actor self, Order order)
+		{
+			if (order.OrderString == "Dock" && CanDockAt(order.Target.Actor, false, true))
+				return Info.Voice;
+			else if (order.OrderString == "ForceDock" && CanDockAt(order.Target.Actor, false, true))
```

Comment:
Could shorten this to `order.OrderString == "Dock" || order.OrderString == "ForceDock"` and keep only one of the `If` statements, to save a call to `CanDockAt()`.

Reply:
This is a bug. I forgot to make the CanDockAt checks different